### ISSUE TITLE
The difference between Keras SGD and tf.train.MomentumOptimizer

### ISSUE THREAD

**Issue by djshen** [cjlin1/simpleNN#7]
There was a discussion about whether we can replace the `tf.compat.v1.train.MomentumOptimizer` with `tf.keras.optimizers.SGD`. To find the difference, I read the document and trace the source code.

The update rule of [SGD](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/SGD) is:
v<sub>t+1</sub> &larr; &alpha;v<sub>t</sub> - &eta;&nabla;f(&theta;<sub>t</sub>)
&theta;<sub>t+1</sub> &larr; &theta;<sub>t</sub> + v<sub>t+1</sub>

And that of [MomentumOptimizer](https://www.tensorflow.org/api_docs/python/tf/compat/v1/train/MomentumOptimizer) is:
v<sub>t+1</sub> &larr; &alpha;v<sub>t</sub> - &nabla;f(&theta;<sub>t</sub>)
&theta;<sub>t+1</sub> &larr; &theta;<sub>t</sub> + &eta;v<sub>t+1</sub>

The difference is that the learning rate is multiplied at the first step in SGD while it is multiplied at the second step in MomentumOptimizer.

If the learning rate &eta; is a constant, the two formula are mathematically equivalent but there might be some floating point errors. However, if the learning rate is changing, the results will be different. Hope this could answer the question.

BTW, I find that the update rule in the slides is the same as SGD instead of MomentumOptimizer.
